File: LinkedList/MiddleNode.py
# Python program to find middle node of a singly linked list


# Node class
class Node:

    # Constructor to initialise data and next
    def __init__(self, data=None):
        self.data = data
        self.next = None


class SinglyLinkedList:

    # Constructor to initialise head
    def __init__(self):
        self.head = None

    # Function to find middle node of a linked list
    def find_mid(self):
        fast = self.head
        slow = self.head

        # Make fast run twice the speed of slow
        # when fast is at the last of the list
        # slow will be at the mid node
        while fast is not None:
            fast = fast.next
            if fast is not None:
                fast = fast.next
                slow = slow.next

        return slow

    # Function to Insert data at the beginning of the linked list
    def insert_at_beg(self, data):
        node = Node(data)
        node.next = self.head
        self.head = node

    # Function to print the linked list
    def print_data(self):
        current = self.head
        while current is not None:
            print(current.data, '-> ', end='')
            current = current.next
        print('None')

if __name__ == '__main__':
    linked_list = SinglyLinkedList()
    linked_list.insert_at_beg(9)
    linked_list.insert_at_beg(8)
    linked_list.insert_at_beg(7)
    linked_list.insert_at_beg(6)
    linked_list.insert_at_beg(5)
    linked_list.insert_at_beg(4)
    linked_list.insert_at_beg(3)
    linked_list.insert_at_beg(2)
    linked_list.insert_at_beg(1)
    linked_list.print_data()
    # call the find_mid function
    mid = linked_list.find_mid()
    # print the middle node if not None
    if mid is not None:
        print(mid.data)

File: LinkedList/MiddleNode.cpp
// Find the middle node of a singly linked list

#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* next;

        Node() {
            this->next = NULL;
        }

        Node(int data) {
            this->data = data;
            this->next = NULL;
        }
};

class LinkedList {
    public:
        Node* head = NULL;

        void insert_at_beg(int data){
            Node* temp = new Node(data);
            temp->next = this->head;
            this->head = temp;
        }

        void print_data(){
            Node* itr = this->head;

            while(itr != NULL){
                cout << itr->data << " ";
                itr = itr->next;
            }
            cout << endl;
        }
};

Node* middle_node(LinkedList* ll){
    Node *fast = ll->head, *slow = ll->head;
    
    // if list is empty or only 1 node is present
    if(fast == NULL || fast->next == NULL)
        return fast;
    
    // while the double speed pointer doesn't get to the end
    // keep updating both slow and fast pointers
    while(fast != NULL){
        fast = fast->next;

        if(fast != NULL){
            fast = fast->next;
            slow = slow->next;
        }
    }

    return slow;
}

int main() {
    LinkedList* ll = new LinkedList();

    // Create Linked list
    ll->insert_at_beg(7);
    ll->insert_at_beg(6);
    ll->insert_at_beg(5);
    ll->insert_at_beg(4);
    ll->insert_at_beg(3);
    ll->insert_at_beg(2);
    ll->insert_at_beg(1);

    cout<<"Linked List data:\n";
    ll->print_data();
    
    cout << "Middle of the linked list is: " << middle_node(ll)->data << endl;
    
    ll->insert_at_beg(0);

    cout<<"Linked List data:\n";
    ll->print_data();
    
    cout << "Middle of the linked list is: " << middle_node(ll)->data << endl;
    return 0;
}


